### ISSUE TITLE
Backfill: change default block size

### ISSUE THREAD

**Issue by roidelapluie** [prometheus/prometheus#8388]
With the newly introduced backfill command, we are creating 2-hour blocks. For a year it is more than 4000 blocks.

It would be nice if we could either automatically compact them or change the default duration of the blocks (there is a TODO for this in code).

Either via a --block-duration flag or --enable-compations and prometheus compaction flags (max block size etc).

**Comment by roidelapluie** [comment]
cc @aSquare14 FYI

**Comment by codesome** [comment]
+1 for `--block-duration` since a bigger block duration can make backfill faster (since we iterate through the input file for every block). Compactions can be left to Prometheus to take care of and not required to have it in 2 places.

**Comment by brian-brazil** [comment]
Any duration which is not a factor of 2h could cause problems when it comes to compaction later on.

**Comment by roidelapluie** [comment]
Do we have a guarantee about alignment in TSDB?

What happens with block size = 8h and a situation like this, is it possible?

![blocksize](https://user-images.githubusercontent.com/291750/105841420-d5d2ce80-5fd4-11eb-9d1f-cfa224da8f2a.png)

**Comment by brian-brazil** [comment]
That's not possible with Prometheus currently, and I believe would end poorly with the current code.

**Comment by stevekuznetsov** [comment]
@roidelapluie @brian-brazil howdy! I'd love to work on this, but I lack the subject matter expertise to understand this fully. Some questions I'd love some input on:

- is it a hard requirement that block durations are a multiple of two hours?
- if so, would it be more appropriate to expose `--block-duration-multiplier` or something and apply that to the base 2h size?
- what is the result of the conversation above w.r.t. alignment in TSDB? I can't parse the conclusion
- as a user without deep knowledge of TSDB internals, what are my actual needs? Why would I ever *not* choose a block size as large as required to contain all of my data that I'm backfilling? What are the tradeoffs between larger and smaller sizes? How would I go about choosing the correct one?

**Comment by roidelapluie** [comment]
I have done this for the graphite exporter: https://github.com/prometheus/graphite_exporter/pull/145

**Comment by stevekuznetsov** [comment]
@roidelapluie thanks for the link - unfortunately I think it really does not answer most of my questions, especially those on the last line of my last comment. I see you validate the input duration like:

```go
func validateBlockDuration(t int64) bool {
	i, f := math.Modf(float64(t) / float64(tsdb.DefaultBlockDuration))
	if f != 0 {
		return false
	}

	i, f = math.Modf(math.Log2(i))
	return f == 0 && i >= 0
}
```

I have a couple questions:

1. why are you using floating point arithmetic here when the inputs are integer numbers of nanoseconds (`time.Duration`)?
2. am I understanding this correctly in that the only valid block sizes are powers of 2 multiples of the base duration?
3. if so, why even expose a duration flag to the users when there's such a narrow range of possible values it can take? why not expose e.g. the integer we'll exponentiate 2 by? For instance `--block-duration-factor 3` would mean a duration of `2 ** 3 * 2h = 16h`

**Comment by stevekuznetsov** [comment]
Mostly though I think from a user perspective where I am not a Prometheus/TSDB expert, I'm not really sure that I want to tell the tool anything about the block duration. I mostly want to make my backfill fast :)

So, to that end - how would I choose a duration? Why would I ever choose a duration that's shorter than the smallest duration that fully covers my input dataset? What are the drawbacks to this?

**Comment by brian-brazil** [comment]
> is it a hard requirement that block durations are a multiple of two hours?

No, but things will almost certainly go wrong unless you know exactly what you're doing.

> if so, would it be more appropriate to expose --block-duration-multiplier or something and apply that to the base 2h size?

Maybe, however a non-trivial number of users have ignored instruction to not mess around with block sizes and compaction settings (e.g. the majority of Thanos users) so this may also cause problems with them. And you still need to worry about alignment.

> Why would I ever not choose a block size as large as required to contain all of my data that I'm backfilling?

After compaction it can cause one block covering all of that time period, plus every other block overlapping it. Such a mega block is not desirable.

**Comment by brian-brazil** [comment]
> As I understand it, you're saying that other projects / derivative users of TSDB like Thanos are not using the same default block size,

No, Thanos users of Prometheus have different block sizes on disk than a Prometheus with default settings would have. The alignment is generally the same for these, but they'd only ever want 2h blocks.

> Why is a mega block not desirable? 

At the simplest, time-based retention won't work as as long as one part of the potentially years-long block is in range it has to be kept. Conversely with size-based retention, data will be deleted far too soon. You can't compact with such a block, without producing more mega blocks which could leave you with lots of blocks around if you backfill regularly.

**Comment by stevekuznetsov** [comment]
@brian-brazil thanks for all the clarifications! One last question: is it possible after connecting to a TSDB instance to figure out a) what the default block size it has been configured to use and b) how to align the blocks? I can't see any alignment happening in the backfill command right now, is that because the current code doesn't actually talk to the TSDB directly, instead simply writing data to the directory? Is this also why it's required to restart Prometheus after backfilling to get the data fully ingested?

**Comment by roidelapluie** [comment]
1) It is not an information stored, but I would assume 2h. I don't think we should support something different in this tool.

2) The blocks are already aligned in the current code:
https://github.com/prometheus/prometheus/blob/d44b4c5f1da2b7e84ad86c73cfde05ca030595c1/cmd/promtool/backfill.go#L70

**Comment by roidelapluie** [comment]
It should not be required to restart Prometheus with the more recent version, we should see the data within 1 minute.

**Comment by brian-brazil** [comment]
> It is not an information stored, but I would assume 2h. I don't think we should support something different in this tool.

More particularly we don't know what values might have been used previously, but 2h is reasonable to presume.

**Comment by stevekuznetsov** [comment]
Thanks all for the pointers and explanations, I appreciate you taking your time to help me understand. I've tried to expose the option to change the duration (along with a discussion of when this might be appropriate and what the drawbacks might be) in #8919 .

**Comment by stevekuznetsov** [comment]
FWIW with the fix from #8919 and a block duration that spans my entire dataset, a 1min import now takes 500ms.
